### PR TITLE
Make GKE module cluster_name computed attribute

### LINKED ISSUE #1181: ASM module not deferring data source read to apply time
### TL;DR

When GKE cluster and ASM module are created in the same config and GKE module inputs are passed to ASM module, [the data source read](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/774fc12df6cb9f801caf72cd0ac6dfdbe723da46/modules/asm/main.tf#L28-L32) for cluster is not deferred to apply time as project_id, cluster name and location are known at plan time. This results in an error as cluster does not exist yet.

```
 Error: Invalid template interpolation value
│ 
│   on ../../../modules/asm/hub.tf line 21, in resource "google_gke_hub_membership" "membership":
│   21:   membership_id = "${data.google_container_cluster.asm.name}-membership"
│     ├────────────────
│     │ data.google_container_cluster.asm.name is null
│ 
│ The expression result is null. Cannot include a null value in a string template.
╵
╷
│ Error: Invalid template interpolation value
│ 
│   on ../../../modules/asm/hub.tf line 24, in resource "google_gke_hub_membership" "membership":
│   24:       resource_link = "//container.googleapis.com/${data.google_container_cluster.asm.id}"
│     ├────────────────
│     │ data.google_container_cluster.asm.id is null
│ 
│ The expression result is null. Cannot include a null value in a string template.
```

Our current tests did not catch this as a random_string was passed as cluster name deferring the read to apply time 
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/774fc12df6cb9f801caf72cd0ac6dfdbe723da46/test/fixtures/simple_zonal_with_asm/example.tf#L25

Module depends on also throws an error due to gcloud module

```
│ Error: Invalid count argument
│ 
│   on .terraform/modules/example.asm.cpr/main.tf line 57, in resource "random_id" "cache":
│   57:   count = (!local.skip_download) ? 1 : 0
```

We can look into using a computed attribute to defer this to apply time. 

### Expected behavior

_No response_

### Observed behavior

_No response_

### Terraform Configuration

```hcl
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/examples/simple_zonal_with_asm with a static cluster name
```

### Terraform Version

```sh
v1.0.5
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Monkeyanator**
Fixes #1181.

Change to the `simple_zonal_with_asm` module is just to get a passing run in CI with a static cluster name.

### CODE REVIEW

**Comment by morgante** [comment]
Approach looks good, note changes need to be made in `autogen/` then run `make build`.

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_generate_modules
 The modules need to be regenerated. Please run `make_build`. 
 ```bash 
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/outputs.tf /tmp/tmp.45yG5Qm2Ib/workspace/outputs.tf
27,36d26
<   depends_on = [
<     /* Nominally, the cluster name is populated as soon as it is known to Terraform.
<     * However, the cluster may not be in a usable state yet.  Therefore any
<     * resources dependent on the cluster being up will fail to deploy.  With
<     * this explicit dependency, dependent resources can wait for the cluster
<     * to be up.
<     */
<     google_container_cluster.primary,
<     google_container_node_pool.pools,
<   ]
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes
 ```

File: modules/beta-autopilot-public-cluster/outputs.tf [hunk 1]
```diff
@@ -24,6 +24,16 @@ output "cluster_id" {
...
   value       = local.cluster_name
+  depends_on = [
+    /* Nominally, the cluster name is populated as soon as it is known to Terraform.
+    * However, the cluster may not be in a usable state yet.  Therefore any
+    * resources dependent on the cluster being up will fail to deploy.  With
+    * this explicit dependency, dependent resources can wait for the cluster
+    * to be up.
+    */
+    google_container_cluster.primary,
+    google_container_node_pool.pools,
```
Comment (bharathkkb) [inline modules/beta-autopilot-public-cluster/outputs.tf:35]: Autopilot wont have pools, so you will need to add a conditional in autogen. Example
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/9f06ef4562adcc9096fbf2da6041746d3d08c483/autogen/main/outputs.tf.tmpl#L60-L63

Comment (Monkeyanator) [reply modules/beta-autopilot-public-cluster/outputs.tf]: Done, thanks

**Comment by Monkeyanator** [comment]
Hm, looks like even after the changes in this PR, CI fails on the ASM module when using a static cluster name:

```
Step #60 - "converge simple-zonal-with-asm-local":        Error: Invalid template interpolation value
Step #60 - "converge simple-zonal-with-asm-local":        
Step #60 - "converge simple-zonal-with-asm-local":          on ../../../modules/asm/hub.tf line 21, in resource "google_gke_hub_membership" "membership":
Step #60 - "converge simple-zonal-with-asm-local":          21:   membership_id = "${data.google_container_cluster.asm.name}-membership"
Step #60 - "converge simple-zonal-with-asm-local":            ├────────────────
Step #60 - "converge simple-zonal-with-asm-local":            │ data.google_container_cluster.asm.name is null
Step #60 - "converge simple-zonal-with-asm-local":        
Step #60 - "converge simple-zonal-with-asm-local":        The expression result is null. Cannot include a null value in a string
Step #60 - "converge simple-zonal-with-asm-local":        template.
Step #60 - "converge simple-zonal-with-asm-local":        
Step #60 - "converge simple-zonal-with-asm-local":        Error: Invalid template interpolation value
Step #60 - "converge simple-zonal-with-asm-local":        
Step #60 - "converge simple-zonal-with-asm-local":          on ../../../modules/asm/hub.tf line 24, in resource "google_gke_hub_membership" "membership":
Step #60 - "converge simple-zonal-with-asm-local":          24:       resource_link = "//container.googleapis.com/${data.google_container_cluster.asm.id}"
Step #60 - "converge simple-zonal-with-asm-local":            ├────────────────
Step #60 - "converge simple-zonal-with-asm-local":            │ data.google_container_cluster.asm.id is null
Step #60 - "converge simple-zonal-with-asm-local":        
Step #60 - "converge simple-zonal-with-asm-local":        The expression result is null. Cannot include a null value in a string
Step #60 - "converge simple-zonal-with-asm-local":        template.
```

**Comment by morgante** [comment]
@Monkeyanator Instead of using `depends_on` what if you make it the output an implicit reference to the cluster resource?

**Comment by Monkeyanator** [comment]
> Sam Naser Instead of using `depends_on` what if you make it the output an implicit reference to the cluster resource?

I thought it already was since the value is `local.cluster_name = local.cluster_output_name = google_container_cluster.primary.name`?

**Comment by morgante** [comment]
> > Sam Naser Instead of using `depends_on` what if you make it the output an implicit reference to the cluster resource?
> 
> I thought it already was since the value is `local.cluster_name = local.cluster_output_name = google_container_cluster.primary.name`?

You're right, it's kind of confusing why this isn't being deferred. What if we try inlining the `.id` workaround from the other PR? There's no way it could know that before applying.

**Comment by bharathkkb** [comment]
Adding the workaround to the gke module output directly should work. However if we are not depending on nodepools, a zonal cluster can go into an upgrade after `google_container_cluster` is resolved and ASM apply could fail. I am thinking we should derive the cluster name from `google_container_node_pool[*].id` for regular clusters and `google_container_cluster.id` for autopilot clusters.

File: autogen/main/main.tf.tmpl [hunk 2]
```diff
@@ -175,6 +175,13 @@ locals {
   cluster_zones    = sort(local.cluster_output_zones)
 
   cluster_name                               = local.cluster_output_name
+  {% if autopilot_cluster != true %}
+  // node pool ID is in the form project/location/cluster/name
```
Comment (bharathkkb) [inline autogen/main/main.tf.tmpl:179]: Based on logic below it looks like this should be the form?
```suggestion
  // node pool ID is in the form projects/<project-id>/locations/<location>/clusters/<cluster-name>/nodePools/<nodepool-name>
```

File: autogen/main/main.tf.tmpl [hunk 3]
```diff
@@ -175,6 +175,13 @@ locals {
   cluster_zones    = sort(local.cluster_output_zones)
 
   cluster_name                               = local.cluster_output_name
```
Comment (bharathkkb) [inline autogen/main/main.tf.tmpl:177]: Do we use this anymore?

**Review by bharathkkb — APPROVED** [review]
Minor comments, LGTM otherwise

File: autogen/main/main.tf.tmpl [hunk 4]
```diff
@@ -175,6 +175,13 @@ locals {
...
+  {% if autopilot_cluster != true %}
+  // node pool ID is in the form project/location/cluster/name
+  cluster_name_computed                      = element(split("/", element(values(google_container_node_pool.pools)[*].id, 0)), length(split("/", element(values(google_container_node_pool.pools)[*].id, 0))) - 3)
```
[the remark this answers is not in the document]
Comment (Monkeyanator) [reply autogen/main/main.tf.tmpl]: Done.

File: autogen/main/main.tf.tmpl (hunk 3 above)
Comment (Monkeyanator) [reply autogen/main/main.tf.tmpl]: Nope, removed

File: autogen/main/main.tf.tmpl (hunk 2 above)
Comment (Monkeyanator) [reply autogen/main/main.tf.tmpl]: Right thanks, updated

**Comment by bharathkkb** [comment]
Test failing due to https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/1197

**Comment by pramodsetlur** [comment]
Hey all, thanks for the fix and comments. I am pointing to the module's master branch and still see the same issue during the apply. Once it occurs on the apply, the issue is seen on the subsequent plans as well. I am creating the GKE cluster alongside with ASM (per the original issue). Not sure what I am missing. 

```
Error: Invalid template interpolation value

  on .terraform/modules/gke_cluster.asm_managed_control_plane/modules/asm/hub.tf line 21, in resource "google_gke_hub_membership" "membership":
  21:   membership_id = "${data.google_container_cluster.asm.name}-membership"
    |----------------
    | data.google_container_cluster.asm.name is null

The expression result is null. Cannot include a null value in a string
template.


Error: Invalid template interpolation value

  on .terraform/modules/gke_cluster.asm_managed_control_plane/modules/asm/hub.tf line 24, in resource "google_gke_hub_membership" "membership":
  24:       resource_link = "//container.googleapis.com/${data.google_container_cluster.asm.id}"
    |----------------
    | data.google_container_cluster.asm.id is null

The expression result is null. Cannot include a null value in a string
template.
```

Here is the reference to the module:
```
module "asm_managed_control_plane" {
  count                     = var.asm_managed_control_plane_enabled == true ? 1 : 0
  source                    = "github.com/terraform-google-modules/terraform-google-kubernetes-engine//modules/asm"
  project_id                = var.project_id
  cluster_name              = "${google_container_cluster.gke_cluster.id}"
  cluster_location          = var.region
  enable_cni                = false
  enable_fleet_registration = true
  enable_mesh_feature       = false
}
```
A few points:
* The `cluster_name              = "${google_container_cluster.gke_cluster.id}"` is not being picked up by data resource even though the cluster seems to have been created. Not sure why it is null. 
* I am using a variable `asm_managed_control_plane_enabled` and install Managed ASM Control plane only if necessary. 
* I had to update the `source` to `github.com/terraform-google-modules/terraform-google-kubernetes-engine//modules/asm` instead of `terraform-google-modules/kubernetes-engine/google//modules/asm` (`terraform-google-kubernetes-engine` vs `kubernetes-engine`) .
* `enable_mesh_feature       = false` - Setting this to true will fail on apply if the `mesh` feature is already enabled (from a previous installation of service mesh). Wondering if it can be made idempotent, so that the code and resource can match.

Any help is appreciated. Thanks in advance!

**Comment by pramodsetlur** [comment]
Hey all, thanks for the fix and comments. I am pointing to the module's master branch and still see the same issue during the apply. Once it occurs on the apply, the issue is seen on the subsequent plans as well. I am creating the GKE cluster alongside with ASM (per the original issue). Not sure what I am missing. 

```
Error: Invalid template interpolation value

  on .terraform/modules/gke_cluster.asm_managed_control_plane/modules/asm/hub.tf line 21, in resource "google_gke_hub_membership" "membership":
  21:   membership_id = "${data.google_container_cluster.asm.name}-membership"
    |----------------
    | data.google_container_cluster.asm.name is null

The expression result is null. Cannot include a null value in a string
template.


Error: Invalid template interpolation value

  on .terraform/modules/gke_cluster.asm_managed_control_plane/modules/asm/hub.tf line 24, in resource "google_gke_hub_membership" "membership":
  24:       resource_link = "//container.googleapis.com/${data.google_container_cluster.asm.id}"
    |----------------
    | data.google_container_cluster.asm.id is null

The expression result is null. Cannot include a null value in a string
template.
```

Here is the reference to the module:
```
module "asm_managed_control_plane" {
  count                     = var.asm_managed_control_plane_enabled == true ? 1 : 0
  source                    = "github.com/terraform-google-modules/terraform-google-kubernetes-engine//modules/asm"
  project_id                = var.project_id
  cluster_name              = "${google_container_cluster.gke_cluster.id}"
  cluster_location          = var.region
  enable_cni                = false
  enable_fleet_registration = true
  enable_mesh_feature       = false
}
```
A few points:
* The `cluster_name              = "${google_container_cluster.gke_cluster.id}"` is not being picked up by data resource even though the cluster seems to have been created. Not sure why it is null. 
* I am using a variable `asm_managed_control_plane_enabled` and install Managed ASM Control plane only if necessary. 
* I had to update the `source` to `github.com/terraform-google-modules/terraform-google-kubernetes-engine//modules/asm` instead of `terraform-google-modules/kubernetes-engine/google//modules/asm` (`terraform-google-kubernetes-engine` vs `kubernetes-engine`) .
* `enable_mesh_feature       = false` - Setting this to true will fail on apply if the `mesh` feature is already enabled (from a previous installation of service mesh). Wondering if it can be made idempotent, so that the code and resource can match.

Any help is appreciated. Thanks in advance!

**Comment by morgante** [comment]
@pramodsetlur Please open a new issue.

**Comment by pramodsetlur** [comment]
Hmm, as I was writing a new issue, i tried explicitly depending on the gke cluster resource - which seem to have worked.

```
module "asm_managed_control_plane" {
  depends_on = [
    google_container_cluster.gke_cluster
  ]
  count = var.asm_managed_control_plane_enabled == true ? 1 : 0
  source                    = "github.com/terraform-google-modules/terraform-google-kubernetes-engine//modules/asm"
  project_id                = var.project_id
  cluster_name              = var.cluster_name
  cluster_location          = var.region
  enable_cni                = false
  enable_fleet_registration = true
  # TODO: Mesh feature already enabled in infrak8s-dev project. To be variablized. 
  enable_mesh_feature       = false
}
```

Thanks for the response.

**Comment by morgante** [comment]
Glad to hear it!

> enable_mesh_feature = false - Setting this to true will fail on apply if the mesh feature is already enabled (from a previous installation of service mesh). Wondering if it can be made idempotent, so that the code and resource can match.

Unfortunately no, this is not possible. The mesh API fails if you try to activate an existing feature that is already enabled.

**Pull request merged**